### PR TITLE
rgw/sts: code for returning an error when an IAM policy

### PR DESCRIPTION
**Pull request by pritha-srivastava**
resource belongs to someone else's tenant.

While parsing the policy it discards the resource element,
but then when an operation is evaluated, since the resource element
is empty, it doesnt evaluate the resource at all and the policy
ends up erroneously allowing actions on resources in other tenants.

Fixes: https://tracker.ceph.com/issues/51219

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by pritha-srivastava** [comment]
This can be fixed in two ways:
1. Return error while attaching the policy - this is easy and this is what I have done in this PR
OR
2. Allow the policy to be attached, dont drop the resource even it is a resource belonging to another tenant, and while evaluating the policy, the input resource will be matched with the resource in the policy and Deny will be returned

We cant check for resource element being empty and check if input resource is not empty and then return Deny, since in a bucket policy resource element is optional and if the admin chooses not to pass a resource then policy evaluation will return a Deny when it shouldn't.

**Review by mattbenjamin — APPROVED** [review]
lgtm

**Comment by pritha-srivastava** [comment]
Upon further investigation, I see here: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_resource.html, that 'Statements must include either a Resource or a NotResource element.' and all examples of bucket policies and role policies that I have seen include a 'Resource' element, so if the policy parser drops the Resource element, because of it being not in proper format or it having a tenant other that one it is supposed to have, then we can check for resource element being empty and check if input resource is not empty and then return Deny (contrary to what I had said before). So I think in addition to flagging the error while setting a policy, we can do the check for already saved policies (bucket or role policies). What do you think @adamemerson @mattbenjamin

**Review by adamemerson — APPROVED** [review]
Looks good to me.

**Comment by cbodley** [comment]
@pritha-srivastava can you please rebase? this is approved but hasn't been QAd

**Comment by cbodley** [comment]
@mattbenjamin @adamemerson when you approve a PR, please add the needs-qa tag so it doesn't get stalled like this!

**Comment by pritha-srivastava** [comment]
> @pritha-srivastava can you please rebase? this is approved but hasn't been QAd

done.

File: src/rgw/rgw_common.cc [hunk 1]
```diff
@@ -1053,12 +1053,12 @@ Effect eval_or_pass(const boost::optional<Policy>& policy,
 		    const rgw::IAM::Environment& env,
 		    boost::optional<const rgw::auth::Identity&> id,
 		    const uint64_t op,
-		    const ARN& arn,
```
Comment (mattbenjamin) [inline src/rgw/rgw_common.cc:1056]: why call it res?

Comment (mattbenjamin) [reply src/rgw/rgw_common.cc:1056]: else where in a lot of coded "res" might stand for "result"--is there another word that can be resource?

**Comment by mattbenjamin** [comment]
@pritha-srivastava (or someone) could we get an rgw suite run?

**Comment by pritha-srivastava** [comment]
> @pritha-srivastava (or someone) could we get an rgw suite run?

https://pulpito.ceph.com/prsrivas-2021-09-23_07:12:21-rgw-wip-rgw-sts-51219-distro-basic-smithi/

**Comment by pritha-srivastava** [comment]
1. assert rgwlog['bucket'].find(bucket_name) == 0 or rgwlog['bucket'].find(bucket_name2) == 0 - I see this error, but this is present in other teuthology runs also
2. cluster [WRN] slow request osd_op(client.4587.0:7259 56.d 56:b65f3fbe:::34d68555-e247-4cec-af18-315a6bc2e09e.4593.56__multipart_key1. - this doesnt look related to this PR
3. cluster [WRN] Health check failed: Degraded data redundancy: 2/4 objects degraded (50.000%), 1 pg degraded (PG_DEGRADED)" in cluster log  - saw several of these, they are unrelated to this PR
4. 2021-09-23T09:20:51.439+0000 7fe1f7082700 -1 ERROR: can't load client cert from /home/ubuntu/cephtest/ca/kmip-client.crt
2021-09-23T09:20:51.440 INFO:tasks.rgw.client.0.smithi039.stdout:2021-09-23T09:20:51.439+0000 7fe1f7082700 -1 140608488875776:error:0200100D:system library:fopen:Permission denied:crypto/bio/bss_file.c:288:fopen('/home/ubuntu/cephtest/ca/kmip-client.crt','r') - this is also not related to this PR.
5.  All the other failures are the regular ones, that have always been there.

https://pulpito.ceph.com/prsrivas-2021-09-20_08:10:25-rgw:sts-wip-rgw-sts-51219-distro-basic-smithi/ - STS tests have passed.

**Comment by pritha-srivastava** [comment]
According to me, this PR is ready to be merged

**Pull request merged**